state/applicationofferuser.go

```diff
@@ -61,29 +61,124 @@ func (st *State) UpdateOfferAccess(offer names.ApplicationOfferTag, user names.U
 	if err != nil {
 		return errors.Trace(err)
 	}
-	op := updatePermissionOp(applicationOfferKey(offerUUID), userGlobalKey(userAccessID(user)), access)
 
-	err = st.db().RunTransaction([]txn.Op{op})
-	if err == txn.ErrAborted {
-		return errors.NotFoundf("existing permissions")
+	buildTxn := func(int) ([]txn.Op, error) {
+		_, err := st.GetOfferAccess(offerUUID, user)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		isAdmin, err := st.isControllerOrModelAdmin(user)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops := []txn.Op{updatePermissionOp(applicationOfferKey(offerUUID), userGlobalKey(userAccessID(user)), access)}
+		if !isAdmin && access != permission.ConsumeAccess && access != permission.AdminAccess {
+			statusOps, err := st.suspendRevokedRelationsOps(offerUUID, user.Id())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, statusOps...)
+		}
+		return ops, nil
 	}
+
+	err = st.db().Run(buildTxn)
 	return errors.Trace(err)
 }
 
+// suspendRevokedRelationsOps suspends any relations the given user has against
+// the specified offer.
+func (st *State) suspendRevokedRelationsOps(offerUUID, userId string) ([]txn.Op, error) {
+	conns, err := st.OfferConnections(offerUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var ops []txn.Op
+	relIdsToSuspend := make(set.Ints)
+	for _, oc := range conns {
+		if oc.UserName() == userId {
+			rel, err := st.Relation(oc.RelationId())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if rel.Suspended() {
+				continue
+			}
+			relIdsToSuspend.Add(rel.Id())
+			suspendOp := txn.Op{
+				C:      relationsC,
+				Id:     rel.doc.DocID,
+				Assert: txn.DocExists,
+				Update: bson.D{{"$set", bson.D{{"suspended", true}}}},
+			}
+			ops = append(ops, suspendOp)
+		}
+	}
+
+	// Add asserts that the relations against the offered application don't change.
+	// This is broader that what we need but's all that's possible.
+	ao := NewApplicationOffers(st)
+	offer, err := ao.ApplicationOfferForUUID(offerUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	app, err := st.Application(offer.ApplicationName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	relations, err := app.Relations()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sameRelCount := bson.D{{"relationcount", len(relations)}}
+	ops = append(ops, txn.Op{
+		C:      applicationsC,
+		Id:     app.doc.DocID,
+		Assert: sameRelCount,
+	})
+	// Ensure any relations not being updated remain alive.
+	for _, r := range relations {
+		if relIdsToSuspend.Contains(r.Id()) {
+			continue
+		}
+		ops = append(ops, txn.Op{
+			C:      relationsC,
+			Id:     r.doc.DocID,
+			Assert: isAliveDoc,
```

Comment:
they might not actually be alive, right? I think just use txn.DocExists?

Reply:
done